### PR TITLE
SD Drive, refactoring, and various GUI improvements

### PR DESCRIPTION
**Pull request by yatli**
Trying to implement "poor man's +Drive"

### CODE REVIEW

**Comment by yatli** [comment]
@jmamma noticed a problem in `FileBrowserPage` derivatives, `..` doesn't change to parent folder.

**Comment by yatli** [comment]
yeah it kind of works, but it's very slow.
should look for a way to batch read/write.

**Comment by yatli** [comment]
~~it takes more than one second to pull a pattern from MD?~~
~~compared to it, sdcard IO is fast.~~

edit: setting turbo to 8x reduces time to ~0.3s, which I think is acceptable.
doing a progress bar now.

**Comment by yatli** [comment]
it appears functional now; before testing, make sure to backup your data :)

**Comment by yatli** [comment]
during a (bad?) load (MC->MD), I accidently corrupted MD states -- OS lockup, not bootable.
Doing a factory reset correctly recovers the unit.

**Comment by yatli** [comment]
tested working.

1. Factory reset, save the snapshot to 0.snp
2. Use C6 to load MD808 and MD909.
3. Save snapshot to MD-8089.snp
4. Load snapshot 0.snp, check the kits and patterns are factory default.
5. Load MD-8089.snp, check the kits and patterns are MD-808/MD-909.

**Comment by jmamma** [comment]
> during a (bad?) load (MC->MD), I accidently corrupted MD states -- OS lockup, not bootable.
> Doing a factory reset correctly recovers the unit.

MD does not do much in the way of checking validity of received Kit + Pattern dumps. 
Might be a useful method of exploiting memory access in the future ;-)

**Comment by jmamma** [comment]
> @jmamma noticed a problem in `FileBrowserPage` derivatives, `..` doesn't change to parent folder.

Did you resolve this?

--

I've been meaning to revisit the filebrowser code. Some of the implementation is a little hacky.

**Comment by yatli** [comment]
> Did you resolve this?

Yes, please see the last commit above.
The code is still in `SDDrivePage.cpp`, I can consolidate things a bit.

Edit: done. All the file pages now use a single implementation.

**Comment by yatli** [comment]
The alert method is re-implemented:

infobox

File: avr/cores/megacommand/MCL/SDDrivePage.cpp [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+  show_dirs = true;
+  strcpy(title, "SD-Drive");
+  FileBrowserPage::init();
+}
+
+void SDDrivePage::save_snapshot() {
+  DEBUG_PRINT_FN();
+
+  MDSound sound;
+  char sound_name[] = "________";
```
Comment (jmamma) [inline avr/cores/megacommand/MCL/SDDrivePage.cpp]: Change this to white space "       ".
The TextInputPage has code to remove trailing white space from file names.

**Comment by jmamma** [comment]
Great work here. Really like the new popup menus.

The TextInputPage feels a little laggy compared to quick select. Didn't see any obvious issue with your rendering routine. Wondering why  🤔

**Comment by yatli** [comment]
yeah, I've also noticed this. when I remove the clear popup content code it becomes snappy, but there's rendering artifacts, wondering why

**Comment by jmamma** [comment]
Will take another look at the code after my late dinner.

One quick point. For basic style consistency. Headings should be in capital letters.

**Comment by yatli** [comment]
> For basic style consistency. Headings should be in capital letters

member functions?

**Comment by yatli** [comment]
also going out for dinner, dev pause :)

**Comment by jmamma** [comment]
Ha. I should have been more verbose. GUI style consistency. Menu headings, should be in capital letters. 

"New Project" == "NEW PROJECT"

File: avr/cores/megacommand/MCL/SDDrivePage.cpp (hunk 1 above)
Comment (yatli) [reply avr/cores/megacommand/MCL/SDDrivePage.cpp]: done

**Comment by yatli** [comment]
This thread is going to be overloaded with too much stuff.
I'll leave the file browser TODOs (copy/paste, rename, delete) in another PR. :)

**Review by yatli — COMMENTED** [review]
went over the code. 
Summary:
- SD drive for reading/writing .snp dumps
- GUI code consolidation (MCLGUI.cpp)
- FileBrowserPage and derivatives revamp
- TextInputPage revamp with quick select pane
- Improved popup boxes
- A progress bar popup
- some (not very relevant) refactoring: 
  - `DEBUG_DUMP(expr(x))` gives `expr(x) = _value_of_x_`.
  - `DEBUG_PRINTLN_FN` now gives the full function names with class name.
  - General warning fixes. Use `const char*` when the content is not supposed to be modified.

**Comment by yatli** [comment]
@jmamma the design sketches are in `avr/core/megacommand/Design`. I'll throw more of these in.

**Comment by yatli** [comment]
@jmamma recommend merge #31 first. I'll clean up the conflicts on my side.

**Comment by jmamma** [comment]
> @jmamma recommend merge #31 first. I'll clean up the conflicts on my side.

I've merged RAMPage into dev

**Comment by yatli** [comment]
merged `dev`, no conflicts. ~~I think this is also ready for merge.~~

I'm having build issues:

```

C:\Users\Yatao\Documents\Arduino\hardware\MIDICtrl20_MegaCommand\avr\cores\megacommand\MCL\RAMPage.cpp:17:29: error: 'NUM_RAM_PAGES' was not declared in this scope
 uint8_t RAMPage::rec_states[NUM_RAM_PAGES];
                             ^

C:\Users\Yatao\Documents\Arduino\hardware\MIDICtrl20_MegaCommand\avr\cores\megacommand\MCL\RAMPage.cpp:18:30: error: 'NUM_RAM_PAGES' was not declared in this scope
 uint8_t RAMPage::slice_modes[NUM_RAM_PAGES];
                              ^

C:\Users\Yatao\Documents\Arduino\hardware\MIDICtrl20_MegaCommand\avr\cores\megacommand\MCL\RAMPage.cpp: In member function 'void RAMPage::setup_ram_rec(uint8_t, uint8_t, uint8_t, uint8_t, uint8_t, uint8_t, uint8_t, uint8_t)':
C:\Users\Yatao\Documents\Arduino\hardware\MIDICtrl20_MegaCommand\avr\cores\megacommand\MCL\RAMPage.cpp:73:3: error: 'rec_states' is not a member of 'RAMPage'
   RAMPage::rec_states[page_id] = STATE_QUEUE;
   ^

C:\Users\Yatao\Documents\Arduino\hardware\MIDICtrl20_MegaCommand\avr\cores\megacommand\MCL\RAMPage.cpp:75:5: error: 'rec_states' is not a member of 'RAMPage'
     RAMPage::rec_states[0] = RAMPage::rec_states[1] = STATE_QUEUE;
     ^

C:\Users\Yatao\Documents\Arduino\hardware\MIDICtrl20_MegaCommand\avr\cores\megacommand\MCL\RAMPage.cpp:75:30: error: 'rec_states' is not a member of 'RAMPage'
     RAMPage::rec_states[0] = RAMPage::rec_states[1] = STATE_QUEUE;
                              ^

C:\Users\Yatao\Documents\Arduino\hardware\MIDICtrl20_MegaCommand\avr\cores\megacommand\MCL\RAMPage.cpp: In member function 'void RAMPage::reverse(uint8_t)':
C:\Users\Yatao\Documents\Arduino\hardware\MIDICtrl20_MegaCommand\avr\cores\megacommand\MCL\RAMPage.cpp:174:7: error: 'slice_modes' is not a member of 'RAMPage'
   if (RAMPage::slice_modes[page_id] == 0) {
       ^

C:\Users\Yatao\Documents\Arduino\hardware\MIDICtrl20_MegaCommand\avr\cores\megacommand\MCL\RAMPage.cpp:178:7: error: 'slice_modes' is not a member of 'RAMPage'
   if (RAMPage::slice_modes[page_id] == 1) {
       ^

C:\Users\Yatao\Documents\Arduino\hardware\MIDICtrl20_MegaCommand\avr\cores\megacommand\MCL\RAMPage.cpp: In member function 'bool RAMPage::slice(uint8_t, uint8_t)':
C:\Users\Yatao\Documents\Arduino\hardware\MIDICtrl20_MegaCommand\avr\cores\megacommand\MCL\RAMPage.cpp:215:16: error: 'slice_modes' is not a member of 'RAMPage'
     } else if (RAMPage::slice_modes[page_id] == 0) {
                ^

C:\Users\Yatao\Documents\Arduino\hardware\MIDICtrl20_MegaCommand\avr\cores\megacommand\MCL\RAMPage.cpp: In member function 'void RAMPage::setup_ram_play(uint8_t, uint8_t, uint8_t, uint8_t)':
C:\Users\Yatao\Documents\Arduino\hardware\MIDICtrl20_MegaCommand\avr\cores\megacommand\MCL\RAMPage.cpp:303:3: error: 'rec_states' is not a member of 'RAMPage'
   RAMPage::rec_states[page_id] = STATE_QUEUE;
   ^

C:\Users\Yatao\Documents\Arduino\hardware\MIDICtrl20_MegaCommand\avr\cores\megacommand\MCL\RAMPage.cpp:305:5: error: 'rec_states' is not a member of 'RAMPage'
     RAMPage::rec_states[0] = RAMPage::rec_states[1] = STATE_QUEUE;
     ^

C:\Users\Yatao\Documents\Arduino\hardware\MIDICtrl20_MegaCommand\avr\cores\megacommand\MCL\RAMPage.cpp:305:30: error: 'rec_states' is not a member of 'RAMPage'
     RAMPage::rec_states[0] = RAMPage::rec_states[1] = STATE_QUEUE;
                              ^

C:\Users\Yatao\Documents\Arduino\hardware\MIDICtrl20_MegaCommand\avr\cores\megacommand\MCL\RAMPage.cpp: In member function 'void RAMPage::setup_ram_play_mono(uint8_t)':
C:\Users\Yatao\Documents\Arduino\hardware\MIDICtrl20_MegaCommand\avr\cores\megacommand\MCL\RAMPage.cpp:379:3: error: 'slice_modes' is not a member of 'RAMPage'
   RAMPage::slice_modes[page_id] = 0;
   ^

C:\Users\Yatao\Documents\Arduino\hardware\MIDICtrl20_MegaCommand\avr\cores\megacommand\MCL\RAMPage.cpp: In member function 'void RAMPage::setup_ram_play_stereo(uint8_t)':
C:\Users\Yatao\Documents\Arduino\hardware\MIDICtrl20_MegaCommand\avr\cores\megacommand\MCL\RAMPage.cpp:392:3: error: 'slice_modes' is not a member of 'RAMPage'
   RAMPage::slice_modes[page_id] = 0;
   ^

C:\Users\Yatao\Documents\Arduino\hardware\MIDICtrl20_MegaCommand\avr\cores\megacommand\MCL\RAMPage.cpp: In member function 'virtual void RAMPage::loop()':
C:\Users\Yatao\Documents\Arduino\hardware\MIDICtrl20_MegaCommand\avr\cores\megacommand\MCL\RAMPage.cpp:442:10: error: 'rec_states' is not a member of 'RAMPage'
     if ((RAMPage::rec_states[page_id] == STATE_QUEUE) &&
          ^

C:\Users\Yatao\Documents\Arduino\hardware\MIDICtrl20_MegaCommand\avr\cores\megacommand\MCL\RAMPage.cpp:445:9: error: 'rec_states' is not a member of 'RAMPage'
         RAMPage::rec_states[0] = RAMPage::rec_states[1] = STATE_RECORD;
         ^

C:\Users\Yatao\Documents\Arduino\hardware\MIDICtrl20_MegaCommand\avr\cores\megacommand\MCL\RAMPage.cpp:445:34: error: 'rec_states' is not a member of 'RAMPage'
         RAMPage::rec_states[0] = RAMPage::rec_states[1] = STATE_RECORD;
                                  ^

C:\Users\Yatao\Documents\Arduino\hardware\MIDICtrl20_MegaCommand\avr\cores\megacommand\MCL\RAMPage.cpp:447:9: error: 'rec_states' is not a member of 'RAMPage'
         RAMPage::rec_states[page_id] = STATE_RECORD;
         ^

C:\Users\Yatao\Documents\Arduino\hardware\MIDICtrl20_MegaCommand\avr\cores\megacommand\MCL\RAMPage.cpp:457:7: error: 'rec_states' is not a member of 'RAMPage'
       RAMPage::rec_states[0] = RAMPage::rec_states[1] = STATE_PLAY;
       ^

C:\Users\Yatao\Documents\Arduino\hardware\MIDICtrl20_MegaCommand\avr\cores\megacommand\MCL\RAMPage.cpp:457:32: error: 'rec_states' is not a member of 'RAMPage'
       RAMPage::rec_states[0] = RAMPage::rec_states[1] = STATE_PLAY;
                                ^

C:\Users\Yatao\Documents\Arduino\hardware\MIDICtrl20_MegaCommand\avr\cores\megacommand\MCL\RAMPage.cpp:459:7: error: 'rec_states' is not a member of 'RAMPage'
       RAMPage::rec_states[page_id] = STATE_PLAY;
       ^

C:\Users\Yatao\Documents\Arduino\hardware\MIDICtrl20_MegaCommand\avr\cores\megacommand\MCL\RAMPage.cpp: In member function 'virtual void RAMPage::display()':
C:\Users\Yatao\Documents\Arduino\hardware\MIDICtrl20_MegaCommand\avr\cores\megacommand\MCL\RAMPage.cpp:545:8: error: 'rec_states' is not a member of 'RAMPage'
   if ((RAMPage::rec_states[page_id] != STATE_NOSTATE)) {
        ^

C:\Users\Yatao\Documents\Arduino\hardware\MIDICtrl20_MegaCommand\avr\cores\megacommand\MCL\RAMPage.cpp:568:11: error: 'rec_states' is not a member of 'RAMPage'
   switch (RAMPage::rec_states[page_id]) {
           ^

C:\Users\Yatao\Documents\Arduino\hardware\MIDICtrl20_MegaCommand\avr\cores\megacommand\MCL\RAMPage.cpp:667:9: error: 'rec_states' is not a member of 'RAMPage'
       ((RAMPage::rec_states[page_id] == STATE_RECORD) ||
         ^

C:\Users\Yatao\Documents\Arduino\hardware\MIDICtrl20_MegaCommand\avr\cores\megacommand\MCL\RAMPage.cpp:668:9: error: 'rec_states' is not a member of 'RAMPage'
        (RAMPage::rec_states[page_id] == STATE_PLAY))) {
         ^

C:\Users\Yatao\Documents\Arduino\hardware\MIDICtrl20_MegaCommand\avr\cores\megacommand\MCL\RAMPage.cpp:669:10: error: 'slice_modes' is not a member of 'RAMPage'
     if ((RAMPage::slice_modes[page_id] == 1) &&
          ^

C:\Users\Yatao\Documents\Arduino\hardware\MIDICtrl20_MegaCommand\avr\cores\megacommand\MCL\RAMPage.cpp:670:10: error: 'rec_states' is not a member of 'RAMPage'
         (RAMPage::rec_states[page_id] != STATE_RECORD)) {
          ^

C:\Users\Yatao\Documents\Arduino\hardware\MIDICtrl20_MegaCommand\avr\cores\megacommand\MCL\RAMPage.cpp: In member function 'virtual bool RAMPage::handleEvent(gui_event_t*)':
C:\Users\Yatao\Documents\Arduino\hardware\MIDICtrl20_MegaCommand\avr\cores\megacommand\MCL\RAMPage.cpp:792:5: error: 'slice_modes' is not a member of 'RAMPage'
     RAMPage::slice_modes[page_id] = 1;
     ^

C:\Users\Yatao\Documents\Arduino\hardware\MIDICtrl20_MegaCommand\avr\cores\megacommand\MCL\RAMPage.cpp:802:5: error: 'slice_modes' is not a member of 'RAMPage'
     RAMPage::slice_modes[page_id] = 0;
     ^
```

**Comment by yatli** [comment]
@jmamma any uncommited changes in `RAMPage.h`?

**Comment by jmamma** [comment]
> @jmamma any uncommited changes in `RAMPage.h`?

Apologies. I've checked-in missing changes and merged to dev

8a9be5d5443a01619e3e6e5c18ab1e7458eee65c

**Comment by jmamma** [comment]
Woops still not compiling. One second.

**Comment by jmamma** [comment]
150d6f1a80ce4af28d5dee0895b46ba3328afa62 compiles now.

**Pull request merged**